core/integration/module_test.go

```diff
@@ -6770,6 +6770,877 @@ func schemaVersion(ctx context.Context) (string, error) {
+					require.Error(t, err)
+					require.ErrorContains(t, err, "no such file or directory")
+				})
+			})
+		}
+	})
+
+	t.Run("deps", func(ctx context.Context, t *testctx.T) {
+		c := connect(ctx, t)
+
+		ctr := goGitBase(t, c).
+			WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
+			WithWorkdir("/work/dep").
+			With(daggerExec("init", "--source=.", "--name=dep", "--sdk=go")).
+			WithNewFile("main.go", `package main
+
+import (
+	"dagger/dep/internal/dagger"
+)
+
+type Dep struct{}
+
+func (m *Dep) GetSource(
+	// +defaultPath="/dep"
+	// +ignore=["!yo"]
+	source *dagger.Directory,
+) *dagger.Directory {
+	return source
+}
+`,
+			).
+			WithNewFile("yo", "yo")
+
+		out, err := ctr.With(daggerCall("get-source", "entries")).Stdout(ctx)
+		require.NoError(t, err)
+		require.Equal(t, "yo\n", out)
+
+		ctr = ctr.
+			WithWorkdir("/work").
+			With(daggerExec("init", "--source=.", "--name=test", "--sdk=go")).
+			With(daggerExec("install", "./dep")).
+			WithNewFile("main.go", `package main
+
+import (
+	"dagger/test/internal/dagger"
+)
+
+type Test struct{}
+
+func (m *Test) GetDepSource() *dagger.Directory {
+	return dag.Dep().GetSource()
+}
+`,
+			)
+
+		out, err = ctr.With(daggerCall("get-dep-source", "entries")).Stdout(ctx)
+		require.NoError(t, err)
+		require.Equal(t, "yo\n", out)
+	})
+
+	t.Run("as module", func(ctx context.Context, t *testctx.T) {
+		c := connect(ctx, t)
+
+		ctr := goGitBase(t, c).
+			WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
+			WithWorkdir("/work/dep").
+			With(daggerExec("init", "--source=.", "--name=dep", "--sdk=go")).
+			WithNewFile("main.go", `package main
+
+import (
+	"dagger/dep/internal/dagger"
+)
+		
+type Dep struct{}
+		
+func (m *Dep) GetSource(
+	// +defaultPath="/dep"
+	// +ignore=["!yo"]
+	source *dagger.Directory,
+) *dagger.Directory {
+	return source
+}
+		`).
+			WithNewFile("yo", "yo")
+
+		ctr = ctr.
+			WithWorkdir("/work").
+			With(daggerExec("init", "--source=.", "--name=test", "--sdk=go")).
+			WithNewFile("main.go", `package main
+
+import (
+	"context"
+
+	"dagger/test/internal/dagger"
+	"github.com/Khan/genqlient/graphql"
+)
+			
+type Test struct{}
+			
+func (m *Test) GetDepSource(ctx context.Context, src *dagger.Directory) (*dagger.Directory, error) {
+	err := src.AsModule().Initialize().Serve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	type DirectoryIDRes struct {
+		Dep struct {
+			GetSource struct {
+				ID string
+			}
+		}
+	}
+
+	directoryIDRes := &DirectoryIDRes{}
+	res := &graphql.Response{Data: directoryIDRes}
+
+	err = dag.GraphQLClient().MakeRequest(ctx, &graphql.Request{
+		Query: "{dep {getSource {id} } }",
+	}, res)
+
+	if err != nil {
+		return nil, err
+	}
+
+
+	return dag.LoadDirectoryFromID(dagger.DirectoryID(directoryIDRes.Dep.GetSource.ID)), nil
+}
+			`,
+			)
+
+		out, err := ctr.With(daggerCall("get-dep-source", "--src", "./dep", "entries")).Stdout(ctx)
+		require.NoError(t, err)
+		require.Equal(t, "yo\n", out)
```

Comment:
I *think* this test is technically passing, but I have a hunch it might be wrong.

So, what if we have a dep module that *doesn't* exist on disk, and we created the directory source inside the `GetDepSource` func? I think it would still read from the disk of the parent.

...I'm not sure that's right, but I'm also not convinced what the right behavior *should* be. Potentially I wonder if we should just not allow a context directory at all in this case (or fallback to `.ContextDirectory`), since there's not really a corresponding client it was loaded from here. cc @sipsma 

No need to block here, since it's a minor point in a relatively unused API (but note, we use it in our CI!).

Reply:
Are you referring to the case where the `Directory.asModule` API is used with a directory that was created with some set of arbitrary API calls?

If so, yeah that wouldn't work atm. The logic for loading context dirs/files is based on the Kind of the module source (e.g. [here](https://github.com/TomChv/dagger/blob/0e3ab2e05c6ef4e28ec6f79b6188bcc12d96ff88/core/modulesource.go#L307)), so I think this just means that modules created from arbitrary directories should be marked as their own separate kind from `Local`/`Git` (need to be marked as a different kind [here](https://github.com/TomChv/dagger/blob/0e3ab2e05c6ef4e28ec6f79b6188bcc12d96ff88/core/schema/module.go#L700)).

Then context can be loaded from them. I think the logic would be almost identical to the logic for loading from git.

Either way, agree this is extremely obscure, doesn't need to be a blocker.

Reply:
> Are you referring to the case where the `Directory.asModule` API is used with a directory that was created with some set of arbitrary API calls?

Exactly this.

> I think this just means that modules created from arbitrary directories should be marked as their own separate kind

I suggested a `ScratchKind` for this back in 
https://github.com/dagger/dagger/pull/7744#discussion_r1670959469, can look at that at some point :tada: